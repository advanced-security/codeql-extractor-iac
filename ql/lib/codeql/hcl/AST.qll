/**
 * Classes and predicates for representing and analyzing HCL (HashiCorp Configuration Language)
 * Abstract Syntax Trees. This module includes all AST node types such as expressions, statements,
 * literals, blocks, function calls, variables, and attributes.
 */

import codeql.hcl.ast.AstNodes
import codeql.hcl.ast.Block
import codeql.hcl.ast.Calls
import codeql.hcl.ast.Literals
import codeql.hcl.ast.Variables
import codeql.hcl.ast.Attributes
