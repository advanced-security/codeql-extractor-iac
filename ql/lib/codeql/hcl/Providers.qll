/**
 * Classes and predicates for analyzing cloud provider resources and configurations in HCL.
 * This module imports provider-specific implementations for major cloud platforms and services
 * commonly used in Terraform and other infrastructure-as-code tools.
 *
 * Supported providers include:
 * - AWS (Amazon Web Services) - EC2, S3, RDS, EKS, and other AWS services
 * - Azure (Microsoft Azure) - Virtual machines, storage accounts, databases, and Azure services
 * - GCP (Google Cloud Platform) - Compute Engine, Cloud Storage, GKE, and other GCP services
 * - GitHub - Repository management, actions, and GitHub-specific resources
 * - Kubernetes - Kubernetes cluster resources and configurations
 * - Helm - Helm chart deployments and package management
 * - OCI (Oracle Cloud Infrastructure) - Oracle cloud services and resources
 * - Alicloud (Alibaba Cloud) - Alibaba cloud platform services
 */

import providers.Alicloud
import providers.AWS
import providers.Azure
import providers.GCP
import providers.GitHub
import providers.Helm
import providers.Kubernetes
import providers.OCI
