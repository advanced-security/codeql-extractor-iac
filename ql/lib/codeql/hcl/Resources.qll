private import codeql.Locations
private import codeql.hcl.AST

// Resources are the most important element in the Terraform language.
// Each resource block describes one or more infrastructure objects, such as
// virtual networks, compute instances, or higher-level components such as DNS
// records.
//
// https://developer.hashicorp.com/terraform/language/resources/syntax
class Resource extends Block {
  Resource() { this.hasType("resource") }

  /**
   * Get the name of the resource.
   */
  string getName() { result = this.getLabel(1) }

  /**
   * Get the provider of the resource.
   */
  string getProvider() { result = "Unknown Provider" }

  /**
   * Returns the resource id.
   */
  string getId() { result = this.getName() }

  /**
   * Returns the resource type.
   */
  string getResourceType() { result = this.getLabel(0) }

  /**
   * Checks to see if the resource type matches the given type.
   */
  predicate hasResourceType(string type) { this.getResourceType() = type }

  override string toString() {
    result = "resource " + this.getResourceType() + " " + this.getName()
  }
}

class Provider extends Block {
  Provider() { this.hasType("provider") }

  string getName() { result = this.getLabel(0) }
}
