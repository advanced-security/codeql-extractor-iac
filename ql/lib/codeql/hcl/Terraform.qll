private import codeql.files.FileSystem
private import codeql.hcl.AST
private import codeql.iac.Dependencies
private import Resources

module Terraform {
  /**
   * A Terraform file.
   */
  class TerraformFile extends File {
    TerraformFile() { this.getBaseName().regexpMatch(".*.(tf|hcl)$") }
  }

  /**
   * A Terraform settings block.
   */
  class Terraform extends Block {
    Terraform() { this.hasType("terraform") }

    /**
     * Get the required provider.
     */
    RequiredProvider getRequiredProvider() {
      result = this.getAttribute("required_providers").getAChild()
    }

    /**
     * Get required version of Terraform.
     */
    string getRequiredVersion() {
      result = this.getAttribute("required_version").(StringLiteral).getValue()
    }
  }

  abstract class RequiredProvider extends Expr {
    /**
     * Gets the name of the provider.
     */
    abstract string getName();

    /**
     * Gets the version of the provider.
     */
    abstract string getVersion();

    /**
     * Gets the semantic version of the provider.
     */
    abstract SemanticVersion getSemanticVersion();

    /**
     * Gets the source of the provider.
     */
    abstract string getSource();
  }

  RequiredProvider getProviderByName(string name) {
    exists(RequiredProvider provider | provider.getName() = name | result = provider)
  }

  /**
   * Basic Terraform required provider String.
   */
  class BasicRequiredProvider extends RequiredProvider, StringLiteral {
    private Terraform terraform;

    BasicRequiredProvider() { this = terraform.getAttribute("required_providers").getAChild() }

    override string toString() { result = "RequiredProvider " + this.getName() }

    override string getName() { result = this.getParent().(Block).getAttributeName(this).getName() }

    override string getVersion() { result = this.getValue() }

    override SemanticVersion getSemanticVersion() { result = this.getValue() }

    /**
     * Basic providers are assumed to be from the Hashicorp namespace.
     */
    override string getSource() { result = "hashicorp/" + this.getName() }
  }

  /**
   * A Terraform required provider object.
   */
  class ComplexRequiredProvider extends RequiredProvider, Object {
    private Terraform terraform;

    ComplexRequiredProvider() { this = terraform.getAttribute("required_providers").getAChild() }

    override string toString() { result = "RequiredProvider " + this.getName() }

    override string getName() { result = this.getParent().(Block).getAttributeName(this).getName() }

    /**
     * Gets the source of the provider.
     */
    override string getSource() {
      result = this.getElementByName("source").(StringLiteral).getValue()
    }

    /**
     * Gets the version of the provider.
     */
    override string getVersion() {
      result = this.getElementByName("version").(StringLiteral).getValue()
    }

    override SemanticVersion getSemanticVersion() { result = this.getVersion() }
  }
}
