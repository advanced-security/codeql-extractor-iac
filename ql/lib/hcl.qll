import codeql.Locations
import codeql.files.FileSystem
import codeql.hcl.ast.AstNodes
import codeql.hcl.Resources
import codeql.hcl.Constants
import codeql.hcl.Locals
import codeql.hcl.Terraform
// providers
import codeql.hcl.Providers
// security
import codeql.hcl.Security
