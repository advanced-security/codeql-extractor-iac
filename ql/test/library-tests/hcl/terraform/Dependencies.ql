import hcl

query predicate dependencies(Dependency d) { any() }

query predicate semver(Dependency d, SemanticVersion v) { d.getSemanticVersion() = v }
