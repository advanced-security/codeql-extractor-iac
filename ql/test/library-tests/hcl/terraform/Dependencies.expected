dependencies
| providers.tf:5:15:8:5 | hashicorp/azurerm@3.0.0 |
| providers.tf:9:14:9:22 | hashicorp/time@0.7.2 |
| providers.tf:10:14:10:22 | hashicorp/random@3.3.1 |
semver
| providers.tf:5:15:8:5 | hashicorp/azurerm@3.0.0 | =3.0.0 |
| providers.tf:9:14:9:22 | hashicorp/time@0.7.2 | ~>0.7.2 |
| providers.tf:10:14:10:22 | hashicorp/random@3.3.1 | ~>3.3.1 |
